File: crates/crystalline_cylinder_analysis/src/fit.rs
//! Robust constrained fitting of the damped-cosine correlation model.

use crate::backend::AnalysisBackend;
use crate::laplace::preferred_coordinate;
use crate::model::{CorrelationSeries, DampedCosineFit};
use crate::PreferredAxis;

const PARAMETER_COUNT: usize = 4;
const PHASE_MARGIN: f64 = 1.0e-6;

/// Numerical controls for bounded soft-L1 fitting.
#[derive(Clone, Copy, Debug)]
pub struct FitConfig {
    pub soft_l1_scale: f64,
    pub tolerance: f64,
    pub maximum_evaluations: usize,
    pub rank_tolerance: f64,
}

/// Evaluate the constrained damped-cosine model.
pub fn damped_cosine(time: &[f64], parameters: &[f64; PARAMETER_COUNT]) -> Vec<f64> {
    let [amplitude, rate, omega, phase] = *parameters;
    let offset = 1.0 - amplitude * phase.cos();
    time.iter()
        .map(|&sample_time| {
            amplitude * (-rate * sample_time).exp() * (omega * sample_time + phase).cos() + offset
        })
        .collect()
}

/// Evaluate the analytic model Jacobian in row-major sample order.
pub fn damped_cosine_jacobian(time: &[f64], parameters: &[f64; PARAMETER_COUNT]) -> Vec<f64> {
    let [amplitude, rate, omega, phase] = *parameters;
    let mut jacobian = Vec::with_capacity(time.len() * PARAMETER_COUNT);
    for &sample_time in time {
        let envelope = (-rate * sample_time).exp();
        let argument = omega * sample_time + phase;
        let (sine, cosine) = argument.sin_cos();
        jacobian.extend_from_slice(&[
            envelope * cosine - phase.cos(),
            -amplitude * sample_time * envelope * cosine,
            -amplitude * sample_time * envelope * sine,
            -amplitude * envelope * sine + amplitude * phase.sin(),
        ]);
    }
    jacobian
}

/// Fit the averaged correlation with multistart bounded least squares.
pub fn fit_damped_cosine<B: AnalysisBackend>(
    backend: &B,
    correlation: &CorrelationSeries,
    omega_grid: &[f64],
    config: FitConfig,
) -> DampedCosineFit {
    validate_inputs(correlation, omega_grid, config);
    let time = &correlation.lag_times;
    let observed = &correlation.pearson_mean;
    let dt = time[1] - time[0];
    let duration = *time.last().expect("correlation is empty");
    let nyquist = std::f64::consts::PI / dt;
    let preferred_omega =
        preferred_coordinate(correlation, PreferredAxis::Omega, omega_grid).coordinate;
    let tail_count = 3.max(observed.len() / 10);
    let mut tail = observed[observed.len() - tail_count..].to_vec();
    tail.sort_by(|left, right| left.total_cmp(right));
    let initial_offset = median(&tail).clamp(-0.5, 0.5);
    let initial_amplitude = (1.0 - initial_offset).clamp(0.05, 2.0);
    let lower = [0.0, 0.0, 0.0, -0.5 * std::f64::consts::PI + PHASE_MARGIN];
    let upper = [
        2.0,
        1.0 / dt,
        nyquist,
        0.5 * std::f64::consts::PI - PHASE_MARGIN,
    ];
    let origin_zero = correlation.origin_counts[0] as f64;
    let weights = correlation
        .origin_counts
        .iter()
        .map(|&count| (count as f64 / origin_zero).sqrt())
        .collect::<Vec<_>>();
    let rate_starts = unique_clamped(
        &[1.0 / duration, 3.0 / duration, 10.0 / duration],
        f64::EPSILON,
        upper[1] * (1.0 - 1.0e-8),
    );
    let omega_starts = unique_clamped(
        &[
            0.5 * preferred_omega,
            preferred_omega,
            1.5 * preferred_omega,
        ],
        f64::EPSILON,
        upper[2] * (1.0 - 1.0e-8),
    );

    let mut evaluations = 0;
    let mut best: Option<FitState> = None;
    for &initial_rate in &rate_starts {
        for &initial_omega in &omega_starts {
            let initial = [initial_amplitude, initial_rate, initial_omega, 0.0];
            let state = optimize_start(
                backend,
                time,
                observed,
                &weights,
                initial,
                lower,
                upper,
                config,
                &mut evaluations,
            );
            if state.converged
                && state.parameters.iter().all(|value| value.is_finite())
                && best
                    .as_ref()
                    .is_none_or(|current| state.objective < current.objective)
            {
                best = Some(state);
            }
        }
    }
    let best = best.expect("damped-cosine fit found no finite solution");
    let prediction = damped_cosine(time, &best.parameters);
    let mean = observed.iter().sum::<f64>() / observed.len() as f64;
    let residual_sum = observed
        .iter()
        .zip(&prediction)
        .map(|(&actual, &fitted)| (actual - fitted).powi(2))
        .sum::<f64>();
    let total_sum = observed
        .iter()
        .map(|&actual| (actual - mean).powi(2))
        .sum::<f64>();
    assert!(total_sum > 0.0, "correlation has no variation");
    let [amplitude, rate, omega, phase] = best.parameters;
    DampedCosineFit {
        amplitude,
        rate,
        omega,
        phase,
        offset: 1.0 - amplitude * phase.cos(),
        r_squared: 1.0 - residual_sum / total_sum,
        evaluations,
        converged: best.converged,
        rate_at_lower_boundary: rate <= (1.0 / duration) * 1.0e-6,
        rate_at_upper_boundary: rate >= upper[1] * (1.0 - 1.0e-4),
        amplitude_at_upper_boundary: amplitude >= upper[0] * (1.0 - 1.0e-4),
        prediction,
    }
}

#[derive(Clone, Copy)]
struct FitState {
    parameters: [f64; PARAMETER_COUNT],
    objective: f64,
    converged: bool,
}

#[allow(clippy::too_many_arguments)]
fn optimize_start<B: AnalysisBackend>(
    backend: &B,
    time: &[f64],
    observed: &[f64],
    weights: &[f64],
    mut parameters: [f64; PARAMETER_COUNT],
    lower: [f64; PARAMETER_COUNT],
    upper: [f64; PARAMETER_COUNT],
    config: FitConfig,
    evaluations: &mut usize,
) -> FitState {
    let mut objective_value = objective(time, observed, weights, &parameters, config.soft_l1_scale);
    *evaluations += 1;
    let mut start_evaluations = 1;
    let mut damping: f64 = 1.0e-3;
    let mut converged = false;
    while start_evaluations < config.maximum_evaluations {
        let prediction = damped_cosine(time, &parameters);
        let jacobian = damped_cosine_jacobian(time, &parameters);
        let rows = time.len();
        let mut normal_matrix = [0.0; PARAMETER_COUNT * PARAMETER_COUNT];
        let mut gradient = [0.0; PARAMETER_COUNT];
        let mut column_norm_squared = [0.0; PARAMETER_COUNT];
        for row in 0..rows {
            let weighted_residual = weights[row] * (prediction[row] - observed[row]);
            let robust_weight =
                (1.0 + (weighted_residual / config.soft_l1_scale).powi(2)).powf(-0.25);
            let residual = robust_weight * weighted_residual;
            let mut jacobian_row = [0.0; PARAMETER_COUNT];
            for column in 0..PARAMETER_COUNT {
                let value = robust_weight * weights[row] * jacobian[row * PARAMETER_COUNT + column];
                jacobian_row[column] = value;
                column_norm_squared[column] += value * value;
                gradient[column] += value * residual;
            }
            for left in 0..PARAMETER_COUNT {
                for right in 0..PARAMETER_COUNT {
                    normal_matrix[left * PARAMETER_COUNT + right] +=
                        jacobian_row[left] * jacobian_row[right];
                }
            }
        }
        let gradient_maximum = projected_gradient_maximum(&parameters, &gradient, &lower, &upper);
        if gradient_maximum <= config.tolerance {
            converged = true;
            break;
        }
        for column in 0..PARAMETER_COUNT {
            normal_matrix[column * PARAMETER_COUNT + column] +=
                damping * column_norm_squared[column].max(1.0e-24);
        }
        let rhs = gradient.map(|value| -value);
        let step = backend.linear_least_squares(
            &normal_matrix,
            PARAMETER_COUNT,
            PARAMETER_COUNT,
            &rhs,
            config.rank_tolerance,
        );
        let mut candidate = parameters;
        for index in 0..PARAMETER_COUNT {
            candidate[index] = (parameters[index] + step[index]).clamp(lower[index], upper[index]);
        }
        let scaled_step = candidate
            .iter()
            .zip(parameters)
            .map(|(&next, current)| (next - current).abs() / (1.0 + current.abs()))
            .fold(0.0, f64::max);
        let candidate_objective =
            objective(time, observed, weights, &candidate, config.soft_l1_scale);
        *evaluations += 1;
        start_evaluations += 1;
        if candidate_objective < objective_value {
            let relative_change =
                (objective_value - candidate_objective) / (1.0 + objective_value.abs());
            parameters = candidate;
            objective_value = candidate_objective;
            damping = (damping * 0.3).max(1.0e-12);
            let secondary_gradient_tolerance = config.tolerance.sqrt();
            if (relative_change <= config.tolerance || scaled_step <= config.tolerance)
                && gradient_maximum <= secondary_gradient_tolerance
            {
                converged = true;
                break;
            }
        } else {
            damping = (damping * 10.0).min(1.0e12);
        }
    }
    FitState {
        parameters,
        objective: objective_value,
        converged,
    }
}

fn projected_gradient_maximum(
    parameters: &[f64; PARAMETER_COUNT],
    gradient: &[f64; PARAMETER_COUNT],
    lower: &[f64; PARAMETER_COUNT],
    upper: &[f64; PARAMETER_COUNT],
) -> f64 {
    (0..PARAMETER_COUNT)
        .map(|index| {
            let at_lower = parameters[index] <= lower[index];
            let at_upper = parameters[index] >= upper[index];
            if (at_lower && gradient[index] > 0.0) || (at_upper && gradient[index] < 0.0) {
                0.0
            } else {
                gradient[index].abs()
            }
        })
        .fold(0.0, f64::max)
}

fn objective(
    time: &[f64],
    observed: &[f64],
    weights: &[f64],
    parameters: &[f64; PARAMETER_COUNT],
    scale: f64,
) -> f64 {
    damped_cosine(time, parameters)
        .into_iter()
        .zip(observed)
        .zip(weights)
        .map(|((prediction, &actual), &weight)| {
            let scaled = weight * (prediction - actual) / scale;
            2.0 * scale * scale * ((1.0 + scaled * scaled).sqrt() - 1.0)
        })
        .sum()
}

fn unique_clamped(values: &[f64], lower: f64, upper: f64) -> Vec<f64> {
    let mut values = values
        .iter()
        .map(|value| value.clamp(lower, upper))
        .collect::<Vec<_>>();
    values.sort_by(|left, right| left.total_cmp(right));
    values.dedup_by(|left, right| left.to_bits() == right.to_bits());
    values
}

fn median(sorted: &[f64]) -> f64 {
    let middle = sorted.len() / 2;
    if sorted.len().is_multiple_of(2) {
        0.5 * (sorted[middle - 1] + sorted[middle])
    } else {
        sorted[middle]
    }
}

fn validate_inputs(correlation: &CorrelationSeries, omega_grid: &[f64], config: FitConfig) {
    let count = correlation.lag_times.len();
    assert!(count >= 4, "fit needs four lag samples");
    assert_eq!(
        correlation.pearson_mean.len(),
        count,
        "correlation lengths differ"
    );
    assert_eq!(
        correlation.origin_counts.len(),
        count,
        "origin counts differ"
    );
    assert_eq!(correlation.lag_times[0], 0.0, "fit must start at zero lag");
    assert!(
        correlation
            .lag_times
            .windows(2)
            .all(|pair| pair[1] > pair[0]),
        "lag times are not increasing"
    );
    let dt = correlation.lag_times[1] - correlation.lag_times[0];
    assert!(
        correlation
            .lag_times
            .windows(2)
            .all(|pair| ((pair[1] - pair[0]) - dt).abs() <= dt.abs().max(1.0) * 1.0e-9),
        "lag spacing is not uniform"
    );
    assert!(omega_grid.len() >= 2, "omega grid needs two points");
    assert!(
        omega_grid
            .iter()
            .all(|value| value.is_finite() && *value > 0.0),
        "fit omega grid must be positive"
    );
    assert!(
        omega_grid.windows(2).all(|pair| pair[1] > pair[0]),
        "omega grid is not increasing"
    );
    assert!(
        config.soft_l1_scale.is_finite() && config.soft_l1_scale > 0.0,
        "soft-L1 scale is invalid"
    );
    assert!(
        config.tolerance.is_finite() && config.tolerance > 0.0,
        "fit tolerance is invalid"
    );
    assert!(
        config.maximum_evaluations >= 2,
        "fit evaluation limit is too small"
    );
    assert!(
        config.rank_tolerance.is_finite() && config.rank_tolerance >= 0.0,
        "rank tolerance is invalid"
    );
}

File: packages/laplacian_analysis/src/fit.zig
//! Bounded multistart soft-L1 fitting of the damped-cosine model.

const std = @import("std");
const dynamics_analysis = @import("dynamics_analysis");
const differentiate = @import("backend/differentiate.zig");
const laplace = @import("laplace.zig");
const linalg = dynamics_analysis.backend.linalg;
const options = @import("options.zig");
const result = @import("result.zig");
const schema = @import("schema.zig");

pub const parameter_count = differentiate.parameter_count;
pub const Parameters = differentiate.Parameters;
const phase_margin = 1.0e-6;

const FitState = struct {
    parameters: Parameters,
    objective: f64,
    converged: bool,
};

const StartValues = struct {
    values: [3]f64 = undefined,
    len: usize = 0,
};

pub fn dampedCosine(
    allocator: std.mem.Allocator,
    time: []const f64,
    parameters: Parameters,
) ![]f64 {
    const values = try allocator.alloc(f64, time.len);
    errdefer allocator.free(values);
    try fillDampedCosine(values, time, parameters);
    return values;
}

pub fn dampedCosineJacobian(
    allocator: std.mem.Allocator,
    time: []const f64,
    parameters: Parameters,
) ![]f64 {
    return differentiate.dampedCosineJacobian(allocator, time, parameters);
}

pub fn fitDampedCosine(
    allocator: std.mem.Allocator,
    context: *dynamics_analysis.backend.Context,
    correlation: dynamics_analysis.CorrelationSeries,
    omega_grid: []const f64,
    fit_options: options.FitOptions,
) !result.DampedCosineFit {
    const spacing = try schema.validateFitInput(correlation, omega_grid, fit_options);
    const time = correlation.lag_times;
    const observed = correlation.pearson;
    const duration = time[time.len - 1];
    const nyquist = std.math.pi / spacing;
    const preferred_omega = (try laplace.preferredCoordinate(
        allocator,
        context,
        correlation,
        .omega,
        omega_grid,
    )).coordinate;

    const tail_count = @max(@as(usize, 3), observed.len / 10);
    const tail = try allocator.dupe(f64, observed[observed.len - tail_count ..]);
    defer allocator.free(tail);
    std.mem.sort(f64, tail, {}, comptime std.sort.asc(f64));
    const initial_offset = clamp(median(tail), -0.5, 0.5);
    const initial_amplitude = clamp(1.0 - initial_offset, 0.05, 2.0);
    const lower = Parameters{ 0.0, 0.0, 0.0, -0.5 * std.math.pi + phase_margin };
    const upper = Parameters{ 2.0, 1.0 / spacing, nyquist, 0.5 * std.math.pi - phase_margin };

    const weights = try allocator.alloc(f64, observed.len);
    defer allocator.free(weights);
    const origin_zero = @as(f64, @floatFromInt(correlation.origin_counts[0]));
    for (correlation.origin_counts, weights) |count, *weight| {
        weight.* = @sqrt(@as(f64, @floatFromInt(count)) / origin_zero);
    }

    const rate_starts = uniqueClamped(
        .{ 1.0 / duration, 3.0 / duration, 10.0 / duration },
        std.math.floatEps(f64),
        upper[1] * (1.0 - 1.0e-8),
    );
    const omega_starts = uniqueClamped(
        .{ 0.5 * preferred_omega, preferred_omega, 1.5 * preferred_omega },
        std.math.floatEps(f64),
        upper[2] * (1.0 - 1.0e-8),
    );

    const prediction_scratch = try allocator.alloc(f64, time.len);
    defer allocator.free(prediction_scratch);
    const jacobian_scratch = try allocator.alloc(f64, try std.math.mul(usize, time.len, parameter_count));
    defer allocator.free(jacobian_scratch);

    var evaluations: usize = 0;
    var best: ?FitState = null;
    var best_finite: ?FitState = null;
    for (rate_starts.values[0..rate_starts.len]) |initial_rate| {
        for (omega_starts.values[0..omega_starts.len]) |initial_omega| {
            const state = try optimizeStart(
                context,
                time,
                observed,
                weights,
                .{ initial_amplitude, initial_rate, initial_omega, 0.0 },
                lower,
                upper,
                fit_options,
                prediction_scratch,
                jacobian_scratch,
                &evaluations,
            );
            if (finiteState(state) and
                (best_finite == null or state.objective < best_finite.?.objective))
            {
                best_finite = state;
            }
            if (state.converged and finiteState(state) and
                (best == null or state.objective < best.?.objective))
            {
                best = state;
            }
        }
    }
    const selected = best orelse best_finite orelse return error.NoFiniteFit;
    const prediction = try dampedCosine(allocator, time, selected.parameters);
    errdefer allocator.free(prediction);

    var mean: f64 = 0.0;
    for (observed) |value| mean += value;
    mean /= @as(f64, @floatFromInt(observed.len));
    var residual_sum: f64 = 0.0;
    var total_sum: f64 = 0.0;
    for (observed, prediction) |actual, fitted| {
        const residual = actual - fitted;
        const centered = actual - mean;
        residual_sum += residual * residual;
        total_sum += centered * centered;
    }
    if (!std.math.isFinite(total_sum) or total_sum <= 0.0) {
        return error.ConstantCorrelation;
    }

    const amplitude = selected.parameters[0];
    const rate = selected.parameters[1];
    const omega = selected.parameters[2];
    const phase = selected.parameters[3];
    return .{
        .amplitude = amplitude,
        .rate = rate,
        .omega = omega,
        .phase = phase,
        .offset = 1.0 - amplitude * @cos(phase),
        .r_squared = 1.0 - residual_sum / total_sum,
        .evaluations = evaluations,
        .converged = selected.converged,
        .rate_at_lower_boundary = rate <= (1.0 / duration) * 1.0e-6,
        .rate_at_upper_boundary = rate >= upper[1] * (1.0 - 1.0e-4),
        .amplitude_at_upper_boundary = amplitude >= upper[0] * (1.0 - 1.0e-4),
        .prediction = prediction,
    };
}

fn optimizeStart(
    context: *dynamics_analysis.backend.Context,
    time: []const f64,
    observed: []const f64,
    weights: []const f64,
    initial: Parameters,
    lower: Parameters,
    upper: Parameters,
    fit_options: options.FitOptions,
    prediction: []f64,
    jacobian: []f64,
    evaluations: *usize,
) !FitState {
    var parameters = initial;
    var objective_value = objective(time, observed, weights, parameters, fit_options.soft_l1_scale);
    evaluations.* += 1;
    var start_evaluations: usize = 1;
    var damping: f64 = 1.0e-3;
    var converged = false;

    while (start_evaluations < fit_options.maximum_evaluations) {
        try fillDampedCosine(prediction, time, parameters);
        try differentiate.fillDampedCosineJacobian(jacobian, time, parameters);
        var gradient_system = try differentiate.assembleSoftL1GradientSystem(
            prediction,
            observed,
            weights,
            jacobian,
            fit_options.soft_l1_scale,
        );
        const gradient_maximum = differentiate.projectedScaledGradientMaximum(
            parameters,
            gradient_system.gradient,
            gradient_system.column_norm_squared,
            lower,
            upper,
        );
        if (gradient_maximum <= fit_options.tolerance) {
            converged = true;
            break;
        }
        for (0..parameter_count) |column| {
            gradient_system.normal_matrix[column * parameter_count + column] +=
                damping * @max(gradient_system.column_norm_squared[column], 1.0e-24);
        }
        var rhs: Parameters = undefined;
        for (gradient_system.gradient, &rhs) |value, *target| target.* = -value;
        const step = try solveStep(
            context,
            gradient_system.normal_matrix,
            rhs,
            fit_options.rank_tolerance,
        );

        var candidate = parameters;
        var scaled_step: f64 = 0.0;
        for (0..parameter_count) |index| {
            candidate[index] = clamp(parameters[index] + step[index], lower[index], upper[index]);
            scaled_step = @max(
                scaled_step,
                @abs(candidate[index] - parameters[index]) / (1.0 + @abs(parameters[index])),
            );
        }
        const candidate_objective = objective(
            time,
            observed,
            weights,
            candidate,
            fit_options.soft_l1_scale,
        );
        evaluations.* += 1;
        start_evaluations += 1;
        const secondary_tolerance = @sqrt(fit_options.tolerance);

        if (candidate_objective < objective_value) {
            const relative_change = (objective_value - candidate_objective) /
                (1.0 + @abs(objective_value));
            parameters = candidate;
            objective_value = candidate_objective;
            damping = @max(damping * 0.3, 1.0e-12);
            if ((relative_change <= fit_options.tolerance or scaled_step <= fit_options.tolerance) and
                gradient_maximum <= secondary_tolerance)
            {
                converged = true;
                break;
            }
        } else {
            if (scaled_step <= fit_options.tolerance and gradient_maximum <= secondary_tolerance) {
                converged = true;
                break;
            }
            damping = @min(damping * 10.0, 1.0e12);
        }
    }
    return .{ .parameters = parameters, .objective = objective_value, .converged = converged };
}

fn solveStep(
    context: *dynamics_analysis.backend.Context,
    normal_matrix: [parameter_count * parameter_count]f64,
    rhs: Parameters,
    rank_tolerance: f64,
) !Parameters {
    var coefficients = try linalg.Matrix(f64).fromHost(
        context,
        parameter_count,
        parameter_count,
        &normal_matrix,
    );
    defer coefficients.deinit();
    var right_hand_side = try linalg.Matrix(f64).fromHost(context, parameter_count, 1, &rhs);
    defer right_hand_side.deinit();
    var solved = try linalg.leastSquares(
        f64,
        coefficients.constView(),
        right_hand_side.constView(),
        .{ .rcond = rank_tolerance },
    );
    defer solved.deinit();
    var step: Parameters = undefined;
    try solved.solution.copyToHost(&step);
    return step;
}

fn fillDampedCosine(output: []f64, time: []const f64, parameters: Parameters) !void {
    if (output.len != time.len) return error.DimensionMismatch;
    try schema.validateModelInput(time, &parameters);
    const amplitude = parameters[0];
    const rate = parameters[1];
    const omega = parameters[2];
    const phase = parameters[3];
    const offset = 1.0 - amplitude * @cos(phase);
    for (time, output) |sample_time, *value| {
        value.* = amplitude * @exp(-rate * sample_time) *
            @cos(omega * sample_time + phase) + offset;
        if (!std.math.isFinite(value.*)) return error.NonFiniteModel;
    }
}

fn objective(
    time: []const f64,
    observed: []const f64,
    weights: []const f64,
    parameters: Parameters,
    scale: f64,
) f64 {
    const amplitude = parameters[0];
    const rate = parameters[1];
    const omega = parameters[2];
    const phase = parameters[3];
    const offset = 1.0 - amplitude * @cos(phase);
    var value: f64 = 0.0;
    for (time, observed, weights) |sample_time, actual, weight| {
        const predicted = amplitude * @exp(-rate * sample_time) *
            @cos(omega * sample_time + phase) + offset;
        const scaled = weight * (predicted - actual) / scale;
        value += 2.0 * scale * scale * (std.math.hypot(1.0, scaled) - 1.0);
    }
    return value;
}

fn finiteState(state: FitState) bool {
    if (!std.math.isFinite(state.objective)) return false;
    for (state.parameters) |value| if (!std.math.isFinite(value)) return false;
    return true;
}

fn uniqueClamped(input: [3]f64, lower: f64, upper: f64) StartValues {
    var output = StartValues{};
    for (input) |raw| {
        const value = clamp(raw, lower, upper);
        var duplicate = false;
        for (output.values[0..output.len]) |existing| {
            if (@as(u64, @bitCast(existing)) == @as(u64, @bitCast(value))) duplicate = true;
        }
        if (!duplicate) {
            output.values[output.len] = value;
            output.len += 1;
        }
    }
    std.mem.sort(f64, output.values[0..output.len], {}, comptime std.sort.asc(f64));
    return output;
}

fn median(sorted: []const f64) f64 {
    const middle = sorted.len / 2;
    return if (sorted.len % 2 == 0)
        0.5 * (sorted[middle - 1] + sorted[middle])
    else
        sorted[middle];
}

fn clamp(value: f64, lower: f64, upper: f64) f64 {
    return @min(@max(value, lower), upper);
}
